File: An 2/Sem 1/MAP/Lab6/C#/Lab2/UI/UI.cs
using System;
using System.Collections;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Lab2.Lab2_Controller;
using Lab2.Lab2_Model;
using Lab2.Lab2_Utils;

namespace Lab2
{
	namespace Lab2_UI {
	    class UI
	    {
	        private Controller controller;

	        private const String kGreeting = "Welcome!";
	        private const String kExitString = "x";
	        private const String kMenuDelimiter = "----------\n";
	        private const String kMenuExit = "(x) Exit";
	        private const String kMenu = "(1) Add student\n"
	                + "(2) Remove students until 10-grade student is found\n"
	                + "(3) All students\n"
	                + "(4) Number of students\n"
					+ "(5) Number of students greater than a given student\n"
	                + kMenuDelimiter
	                + kMenuExit;
	        private const String kStudentAddedMessage = "Student Added!";
	        private const String kStudentsRemovedMessage = "Students Removed!";
	    
	        public UI(Controller controller) {
	            this.controller = controller;
	        }
	    
	        public void showMenu() {
	            System.Console.WriteLine(kGreeting);
	            String userInput = "";
	            while (!kExitString.Equals(userInput)) {
	                System.Console.WriteLine(kMenu);
	                userInput = System.Console.ReadLine();
	                if ("1".Equals(userInput)) {
	                    this.addStudent();
	                } else if ("2".Equals(userInput)) {
	                    this.removeStudents();
	                } else if ("3".Equals(userInput)) {
	                    this.allStudents();
	                } else if ("4".Equals(userInput)) {
	                    this.numberOfStudents();
					} else if ("5".Equals(userInput)) {
						this.numberOfStudentGreaterThan();
					}
	            }
	        }
	    
	        private void addStudent() {
				int id, grade, grade2, grade3, studentType;
				String name, supervisor, thesis;
				ArrayList errorList;
	                
	            try{
	                System.Console.Write("Student ID: ");
	                id = Convert.ToInt16(System.Console.ReadLine());
	                System.Console.Write("Student Name: ");
	                name = System.Console.ReadLine();
	                System.Console.Write("Student Grade: ");
	                grade = Convert.ToInt16(System.Console.ReadLine());

					Console.Write("Student Type:\n(1)Regular\n(2)Graduate\n(3)Undergraduate\n(4)PhD");
					studentType = Convert.ToInt16(System.Console.ReadLine());

					switch (studentType) {
					case 1:
						errorList = this.controller.addStudent(id, name, grade);
						break;
					case 2:
						System.Console.Write("Grade2: ");
						grade2 = Convert.ToInt16(System.Console.ReadLine());
						System.Console.Write("Grade3: ");
						grade3 = Convert.ToInt16(System.Console.ReadLine());
						System.Console.Write("Supervisor: ");
						supervisor = System.Console.ReadLine();
						errorList = this.controller.addStudent(id, name, grade, grade2, grade3, supervisor);
						break;
					case 3:
						System.Console.Write("Grade2: ");
						grade2 = Convert.ToInt16(System.Console.ReadLine());
						errorList = this.controller.addStudent(id, name, grade, grade2);
						break;
					case 4:
						System.Console.Write("Grade2: ");
						grade2 = Convert.ToInt16(System.Console.ReadLine());
						System.Console.Write("Supervisor: ");
						supervisor = System.Console.ReadLine();
						System.Console.Write("Thesis: ");
						thesis = System.Console.ReadLine();

						errorList = this.controller.addStudent(id, name, grade, grade2, supervisor, thesis);
						break;
					default:
						System.Console.Write("Invalid choice!");
						return;
					}
					if (errorList.Count != 0) {
	                    foreach (String error in errorList) {
	                        System.Console.WriteLine(error);
	                    }
	                } else {
	                    System.Console.WriteLine(kStudentAddedMessage);
	                }
	            } catch (Exception) {
	                System.Console.WriteLine("Invalid input!");
	            }
	        
	        }
	    
	        private void removeStudents() {
				try {
	            	this.controller.removeStudentsUntilMaxGrade();
	            	System.Console.WriteLine(kStudentsRemovedMessage);
				}
				catch (StackException e) {
					System.Console.WriteLine (e.exceptionMessage);
				}
	        }
	    
	        private void allStudents() {
				ArrayList allStudents = this.controller.allStudents();
				foreach (String student in allStudents) {
					System.Console.WriteLine(student);
				}
	        }
	    
	        private void numberOfStudents() {
	            System.Console.Write("Total number of students: ");
	            System.Console.Write(this.controller.numberOfStudents());
	            System.Console.WriteLine();
	        }

		private void numberOfStudentGreaterThan() {
			System.Console.Write("Student id: ");
			int id = Convert.ToInt16 (System.Console.ReadLine ());
			int no = this.controller.numberOfStudentGreaterThan(id);
			System.Console.WriteLine("Number of students greater than {0}: {1}\n", id, no);
		}
	    }
	}
}

File: An 2/Sem 1/MAP/Lab9/Java/src/lab8/UI/UI.java
package lab8.UI;

import lab8.Controller.Controller;
import lab8.Utils.StackException;

import java.io.IOException;
import java.util.ArrayList;
import java.util.InputMismatchException;
import java.util.Scanner;

/**
 *
 * @author mihai
 */
public class UI {
    private Controller controller;
    private StudentsObserver lessThanFiveObserver = new StudentsObserver("<");
    private StudentsObserver greaterOrEqualToFiveObserver = new StudentsObserver(">=");

    final String kGreeting = "Welcome!";
    final String kExitString = "x";
    final String kMenuDelimiter = "----------\n";
    final String kMenuExit = "(x) Exit";
    final String kMenu = "(1) Add student\n"
            + "(2) Remove students until 10-grade student is found\n"
            + "(3) All students\n"
            + "(4) Number of students\n"
            + "(5) Save students to file\n"
            + kMenuDelimiter
            + kMenuExit;
    final String kStudentAddedMessage = "Student Added!";
    final String kStudentsRemovedMessage = "Students Removed!";
    
    public UI(Controller controller) {
        this.controller = controller;
        this.controller.addObserver(this.lessThanFiveObserver);
        this.controller.addObserver(this.greaterOrEqualToFiveObserver);
    }
    
    public void showMenu() throws IOException {
        System.out.println(this.kGreeting);
        String userInput = "";
        Scanner input = new Scanner(System.in);
        while (!this.kExitString.equals(userInput)) {
            System.out.println(this.kMenu);
            try {
                userInput = input.nextLine();
                if ("1".equals(userInput)) {
                    this.addStudent();
                } else if ("2".equals(userInput)) {
                    this.removeStudents();
                } else if ("3".equals(userInput)) {
                    this.allStudents();
                } else if ("4".equals(userInput)) {
                    this.numberOfStudents();
                } else if ("5".equals(userInput)) {
                    this.saveStudentsToFile();
                }
            } catch (InputMismatchException e) {
                
            }
   
        }
    }
    
    private void addStudent() {
        Scanner input = new Scanner(System.in);
        input.useDelimiter("\n");
        int id, grade, grade2, grade3, studentType;
        String name, supervisor, thesis;
        ArrayList <String> errorList;
                
        try{
            System.out.print("Student ID: ");
            id = input.nextInt();
            input.nextLine();
            System.out.print("Student Name: ");
            name = input.nextLine();
            System.out.print("Student Grade: ");
            grade = input.nextInt();

            System.out.print("Student Type:\n(1)Regular\n(2)Graduate\n(3)Undergraduate\n(4)PhD");
            studentType = input.nextInt();

            switch (studentType) {
                case 1:
                    errorList = this.controller.addStudent(id, name, grade);
                    break;
                case 2:
                    System.out.print("Grade2: ");
                    grade2 = input.nextInt();
                    input.nextLine();
                    System.out.print("Grade3: ");
                    grade3 = input.nextInt();
                    input.nextLine();
                    System.out.print("Supervisor: ");
                    supervisor = input.nextLine();

                    errorList = this.controller.addStudent(id, name, grade, grade2, grade3, supervisor);
                    break;
                case 3:
                    System.out.print("Grade2: ");
                    grade2 = input.nextInt();

                    errorList = this.controller.addStudent(id, name, grade, grade2);
                    break;
                case 4:
                    System.out.print("Grade2: ");
                    grade2 = input.nextInt();
                    input.nextLine();
                    System.out.print("Supervisor: ");
                    supervisor = input.nextLine();
                    System.out.print("Thesis: ");
                    thesis = input.nextLine();

                    errorList = this.controller.addStudent(id, name, grade, grade2, supervisor, thesis);
                    break;
                default:
                    System.out.print("Invalid choice!");
                    return;
            }
            if (!errorList.isEmpty()) {
                for (String error : errorList) {
                    System.out.println(error);
                }
            } else {
                System.out.println(this.kStudentAddedMessage);
            }
        } catch (InputMismatchException e) {
            System.out.println("Invalid input!");
        }
        
    }
    
    private void removeStudents() {
        try {
            this.controller.removeStudentsUntilMaxGrade();
            System.out.println(this.kStudentsRemovedMessage);
        } catch (StackException ex) {
            System.out.println(ex.getMessage());
        }
    }
    
    private void allStudents() {
        ArrayList<String> allStudents = this.controller.allStudents();
        for (String student : allStudents) {
            System.out.println(student);
        }
    }
    
    private void numberOfStudents() {
        System.out.printf("Total number of students: %d\n", this.controller.numberOfStudents());
    }

    private void saveStudentsToFile() {
        Scanner input = new Scanner(System.in);
        System.out.println("Filename: ");
        String filename = input.nextLine();

        this.controller.saveStudentsToFile(filename);
    }
}
